lib/seattleflu/db/cli/command/etl/consensus_genome.py

```diff
@@ -0,0 +1,363 @@
+
+        LOG.error(f"Aborting with error")
+        raise error from None
+
+    else:
+        processed_without_error = True
+
+    finally:
+        if action == "prompt":
+            ask_to_commit = \
+                "Commit all changes?" if processed_without_error else \
+                "Commit successfully processed consensus genome records up to this point?"
+
+            commit = click.confirm(ask_to_commit)
+        else:
+            commit = action == "commit"
+
+        if commit:
+            LOG.info(
+                "Committing all changes" if processed_without_error else \
+                "Committing successfully processed consensus genome records up to this point")
+            db.commit()
+
+        else:
+            LOG.info("Rolling back all changes; the database will not be modified")
+            db.rollback()
+
+
+def find_sample(db: DatabaseSession, document: dict) -> Any:
+    """
+    Find sample by its record *document* and return sample.
+    Similar to the :func:``find_sample`` from the shared __init__ module, except
+    that this version of the function doesn't lock the sample table for update.
+    """
+    identifier = document['sample_identifier']
+    LOG.debug(f"Looking up sample «{identifier}»")
+
+    sample = db.fetch_row("""
+        select sample_id as id, identifier, encounter_id
+          from warehouse.sample
+         where identifier = %s or
+               collection_identifier = %s
+        """, (identifier,identifier,))
+
+    if not sample:
+        LOG.error(f"No sample with identifier «{identifier}» found")
+        return None
+
+    LOG.info(f"Found sample {sample.id} «{sample.identifier}»")
+    return sample
+
+
+def find_or_create_sequence_read_set(db: DatabaseSession, document: dict, sample: Any) -> Any:
+    """
+    Find sequence read set given a *sample* and consensus genome record
+    *document*, inserting the sequence read set if it does not exist.
+    Return the sequence read set.
+    """
+    urls = document['metadata']['URLs']
+    LOG.debug(dedent(f"""
+    Looking up sequence read set with sample ID «{sample.id}» and URLs {urls}
+    """))
+
+    sequence_read_set = db.fetch_row("""
+        select sequence_read_set_id as id, sample_id, urls
+          from warehouse.sequence_read_set
+         where sample_id = %s
+          and urls @> %s
+          and %s @> urls
+        """, (sample.id, urls, urls))
+
+    if sequence_read_set:
+        LOG.info(f"Found sequence read set {sequence_read_set.id}")
+    else:
+        LOG.debug(dedent(f"""
+        Sequence read set not found for sample id «{sample.id}» and URLs {urls}
+        """))
+
+        data = {
+            "sample_id": sample.id,
+            "urls": urls,
+        }
+
+        sequence_read_set = db.fetch_row("""
+            insert into warehouse.sequence_read_set (sample_id, urls)
+                values (%(sample_id)s, %(urls)s)
+            returning sequence_read_set_id as id
```

Comment:
Also needs to return `sample_id` or else func:`upsert_genome` errors out

Reply:
Thank you, forgot to test that.